### ISSUE TITLE
Supporting archive assets

### ISSUE THREAD

**Issue by viridia** [bevyengine/bevy#12279]
## What problem does this solve or what need does it fill?

From the discord: @cart "Ok I'm reasonably convinced that we need a better answer for "archive style" assets. Maybe we need a per-asset configuration that defines how root asset lifetimes work".

Terms: An "archive" asset in this context is a type of bundle file whose purpose is to contain other assets. Zip files are one example, but this pattern is also often seen in JSON, MessagePack and GLTF files.

There are two kinds of archives: seekable, and non-seekable. Seekable archives are ones where it is possible to load an individual record without reading the entire file serially, by seeking directly to the start of the record. Zip files are an example of this: they contain a directory index which contains the file offset of every entry. I think that this use case is already adequately supported in Bevy, using a custom `AssetReader`, so I don't think we need to do anything.

Non-seekable archives are ones where you have to read the whole thing in order to load any part. JSON, MessagePack, and CBOR are examples. These are currently problematic in Bevy.

Often a developer will want to bundle together related assets to minimize the number of i/o operations: if the assets are all loaded together, then the overhead of opening and reading a file need only be performed once instead of opening a file for each asset individually.

Unfortunately, in current Bevy doing this can actually result in worse performance instead of better. The reason is due to the way asset lifetimes work. The asset system produces a handle for both the container and for each labeled asset, but if you only hold on to the handle for the labeled asset, then the container is immediately dropped. If you then want to subsequently load another labeled asset from the same archive, the asset system has to re-load and re-parse the entire container all over again.

Note that the word "subsequently" in the preceding paragraph has two sub-cases: one is loading two labeled assets from the same container at the same time. The other is loading a labeled asset, and then loading a second labeled asset from the same container at a later time, but while the first handle is still being held.

I'm not sure what exactly is Bevy's behavior in either of these cases: the docs don't say. But in either case, it would be more efficient if we could keep the container asset in memory long enough to retrieve the second asset.

A third possible case is caching, in which a handle is loaded, dropped, and a second handle from the same container is loaded shortly thereafter. However, I think that this case is probably out of scope. Developers can accommodate this case by maintaining an LRU of recently-loaded handles.

Finally, there is the need for keeping the container in memory for purposes of editing. Non-seekable assets cannot be written to disk without writing the whole thing: in order to modify one record, you need to load the entire archive, modify individual records, and then write out the whole thing.

For all of these use cases, the solution is to provide a way to keep the entire container in memory as long as any individual labeled asset is held.

A naive solution to the problem is to design the data type for labeled assets to store a strong handle to the container. Unfortunately, this creates the possibility of a "reference cycle" such that the container will never be freed.

Also, any solution needs to accommodate the case of records which themselves contain dependencies/handles to other records in the same container, or records in other containers: reference cycles should not be created unless the dependencies are actually cyclic.

## What solution would you like?

I'm not going to specify a specific solution here yet, because I want to collect feedback on possible solutions. But the general form of the solution should be a way to keep containers in memory as long as a handle is held. This should be an option in the asset loader implementation.

## What alternative(s) have you considered?

I've played around with storing additional handles in the labeled asset, but I worry about creating reference cycles by doing this.

**Comment by viridia** [comment]
A couple more thoughts on this:

For purposes of editing, it will be useful to be able to iterate through all the entries in the archive. For example, a user may want to select items from a catalog via an interactive chooser ui, which will need a way to iterate the entries.

The obvious way to do this is to have the root asset maintain a hash map of all the entries by label, however it's not clear whether this should be a map of handles or a map of arcs: that is, should the catalog's internal directory go through the asset handle mechanism, or should it hold on to the entries in a more direct way? As mentioned previously, we want to avoid creating cyclic handle dependencies where possible.

Also for purposes of editing, there should be some means of gaining access to the catalog given a handle to an entry; however I think this can be done via manipulation of the asset path - strip off the label portion and then get the handle from the asset server via `.load()`. The reason this is needed is to be able to decide what needs to be saved when an entry is modified.

**Comment by berkus** [comment]
Can we brainstorm the approaches to possible implementation here or it's a purely tracking issue?

**Comment by kristoff3r** [comment]
> Can we brainstorm the approaches to possible implementation here or it's a purely tracking issue?

Yeah discussing solutions is fine. 

I also hit this recently trying to create an asset source for VPK files, which has the added complication that it's usually split into many smaller files that all need some metadata stored to efficiently read assets from it.

**Comment by andriyDev** [comment]
I discussed this idea a bit on [Discord](https://discord.com/channels/691052431525675048/749332104487108618/1430629996438749184). Here's an idea I had.

First, I am not a fan of using the term "archive asset" for this issue. I suggest we split this into two kinds:

- Random-access assets: these are assets where we can efficiently load one subasset contained in the larger asset. For example, a ZIP file contains many assets inside, but a user might just want to load a single one of those assets, and not the whole ZIP file.
- Composite assets: these are assets which contain a root asset and a bunch of subassets that the root asset uses. For example, a GLTF file is composed of many assets, but in total they represent a whole thing.

This issue is primarily discussing the latter. Rather than marking the asset as a composite asset to keep the root asset alive for longer, what if instead we made it so that holding any subasset handle also keeps the root asset and all other subassets alive.

This means that loading a single mesh from a GLTF file will keep all the materials, scenes, etc loaded from the scene. So requesting any of those will not require an additional load. In addition, if we try to load a subasset and we see that its root asset is alive, we can immediately determine whether the subasset exists or not, rather than needing to do another full load just to see that that subasset doesn't exist.

A big risk here is the ZIP case. Users may already be living with these random-access assets by just loading the whole thing and then picking out everything they need, then dropping the unneeded assets. Before we can make subassets keep the root alive, we need to provide an alternative for loading these random-access assets so as not to break users without a workaround.

**Comment by berkus** [comment]
I want to focus on the random access asset case instead. Maybe split this issue?

**Comment by andriyDev** [comment]
I think splitting this issue is a good idea! Happy to chat about it if you file another issue.

Edit: Actually I decided I'll just do this myself haha

**Comment by andriyDev** [comment]
@berkus #21641.
